### PR TITLE
Fix error when deleting old schedule entries.

### PR DESCRIPTION
**Pull request by SpotlightKid**
* Decode static schedule entries keys returned as bytes from Redis.
* Pass app instance to `RedBeatSchedulerEntry`.

```
celery beat v3.1.23 (Cipater) is starting.
__    -    ... __   -        _
Configuration ->
    . broker -> redis://127.0.0.1:6379/1
    . loader -> celery.loaders.app.AppLoader
    . scheduler -> redbeat.schedulers.RedBeatScheduler
       . redis -> redis://127.0.0.1:6379/1
       . lock -> `redbeat::lock` 25.00 seconds (25s)
    . logfile -> [stderr]@%DEBUG
    . maxinterval -> 10.00 seconds (10.0s)
[2016-08-09 19:31:16,576: DEBUG/MainProcess] Setting default socket timeout to 30
[2016-08-09 19:31:16,577: INFO/MainProcess] beat: Starting...
[2016-08-09 19:31:16,581: CRITICAL/MainProcess] beat raised exception <class 'AttributeError'>: AttributeError("'NoneType' object has no attribute 'now'",)
Traceback (most recent call last):
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/kombu/utils/__init__.py", line 323, in __get__
    return obj.__dict__[self.__name__]
KeyError: 'scheduler'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/apps/beat.py", line 112, in start_scheduler
    beat.start()
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/beat.py", line 470, in start
    humanize_seconds(self.scheduler.max_interval))
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/kombu/utils/__init__.py", line 325, in __get__
    value = obj.__dict__[self.__name__] = self.__get(obj)
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/beat.py", line 512, in scheduler
    return self.get_scheduler()
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/beat.py", line 507, in get_scheduler
    lazy=lazy)
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/utils/imports.py", line 53, in instantiate
    return symbol_by_name(name)(*args, **kwargs)
  File "/home/chris/work/celery/redbeat/redbeat/schedulers.py", line 190, in __init__
    super(RedBeatScheduler, self).__init__(app, **kwargs)
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/beat.py", line 185, in __init__
    self.setup_schedule()
  File "/home/chris/work/celery/redbeat/redbeat/schedulers.py", line 199, in setup_schedule
    RedBeatSchedulerEntry(name).delete()
  File "/home/chris/work/celery/redbeat/redbeat/schedulers.py", line 71, in __init__
    args=args, kwargs=kwargs, **clsargs)
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/beat.py", line 94, in __init__
    self.last_run_at = last_run_at or self._default_now()
  File "/home/chris/lib/virtualenvs/celery/lib/python3.5/site-packages/celery/beat.py", line 98, in _default_now
    return self.schedule.now() if self.schedule else self.app.now()
AttributeError: 'NoneType' object has no attribute 'now'
```

---
This change is [Reviewable](https://reviewable.io/reviews/sibson/redbeat/34)

### CODE REVIEW

**Comment by SpotlightKid** [comment]
Error happens when restarting celery beat with redbeat because it looks for old schedule entries in Redis, which returns them as bytes. When comparing them to the current static schedule entries, redbeat thinks they need to be removed, because building a set difference between the Redis schedule keys (bytes) and the current static schedule keys (strings) always yields a non-empty set with the schedule keys as bytes. Redbeat thus tries to remove these schedule entries, which leas to the error cited above, because instantiating a `RedBeatSchedulerEntry` without passing a schedule or app instance is not handled correctly by the super-class `celery.beat.ScheduleEntry`.

On Python 3 the error should happen on every restart, on Python 2 every time you remove a static schedule entry and restart.

**Comment by SpotlightKid** [comment]
> On Python 3 the error should happen on every restart, on Python 2 every time you remove a static schedule entry and restart.

In fact, the error would also happen on Python 2 every time, since the static schedule entry keys weren't remove from the `redbeat::statics'` set  in Redis, as they should be. Added a commit to fix this.

**Comment by sibson** [comment]
going to need to time to reload enough context to review this, feel free to ping me if I don't come back :)

**Comment by SpotlightKid** [comment]
Steps to reproduce:

Python 3

- Add a static schedule entry via `CELERYBEAT_SCHEDULE` setting.
- (If you used redbeat with the same Redis DB before, run `echo FLUSHDB | redis`.)
- Start celery beat with redbeat scheduler.
- Stop celery beat and restart it. Error occurs directly at startup.

Python 2

- Add a static schedule entry via `CELERYBEAT_SCHEDULE` setting.
- (If you used redbeat with the same Redis DB before, run `echo FLUSHDB | redis`.)
- Start celery beat with redbeat scheduler.
- Stop celery beat.
- Remove one entry from `CELERYBEAT_SCHEDULE` setting.
- Restart celery beat. Error occurs directly at startup.

File: redbeat/schedulers.py [hunk 1]
```diff
@@ -191,11 +191,13 @@ def __init__(self, app, **kwargs):
     def setup_schedule(self):
         # cleanup old static entries
         client = redis(self.app)
-        previous = client.smembers(self.app.conf.REDBEAT_STATICS_KEY)
-        current = set(self.app.conf.CELERYBEAT_SCHEDULE.keys())
-        removed = previous - current
+        previous = set(key.decode('utf-8')
```
Comment (sibson) [inline redbeat/schedulers.py]: would it be better to push py3 handling into pyredis rather than handling in app code?  not a blocker on the PR.

File: redbeat/schedulers.py [hunk 2]
```diff
@@ -191,11 +191,13 @@ def __init__(self, app, **kwargs):
...
+                       for key in client.smembers(self.app.conf.REDBEAT_STATICS_KEY))
+        removed = previous.difference(self.app.conf.CELERYBEAT_SCHEDULE.keys())
         for name in removed:
-            RedBeatSchedulerEntry(name).delete()
+            logger.debug("Removing old schedule entry '%s'.", name)
+            RedBeatSchedulerEntry(name, app=self.app).delete()
+            client.srem(self.app.conf.REDBEAT_STATICS_KEY, name)
```
Comment (sibson) [inline redbeat/schedulers.py]: I wonder if we should just client.delete(self.app.conf.REDBEAT_STATICS_KEY) at the start of the loop.  As we're going to rebuild/update all valid keys later.

**Comment by sibson** [comment]
generally +1 on this as is.  Would like to hear your thoughts on the comments before we merge.

File: redbeat/schedulers.py (hunk 1 above)
Comment (SpotlightKid) [inline redbeat/schedulers.py]: In fact, redis-py supports the `decode_reponses` arguments to the `StrictRedis`, which makes it decode all responses from Redis to unicode. My first approach for #33 was to set this to `True` and it generally works, but the tests use `fakeredis`, which does not support `decode_responses` (the [PR for adding it](https://github.com/jamesls/fakeredis/pull/116) doesn't seem likely to be merged anytime soon).

Also, I find it much clearer to explicitly to decode the Redis reponses in each case where it is necessary. Since redbeat has control over what goes into the Redis keys it reads, there should also be no issue with not knowing which encoding to use. Lastly, it makes is more robust in case somebody replaces the default `StrictRedis` instance redbeat uses.

File: redbeat/schedulers.py (hunk 2 above)
Comment (SpotlightKid) [inline redbeat/schedulers.py]: This could leave the schedule information in Redis in a inconsistent state if an exception occurs while deleting a removed static schedule entry. In fact, deleting the schedule entry and the set member should be wrapped in a transaction using `pipeline()` (they can be nested without problems).

I'll update the PR to address this.

**Comment by SpotlightKid** [comment]
I replied to your comments and updated the PR. Feel free to squash the commits if you merge it.

File: redbeat/schedulers.py (hunk 1 above)
Comment (SpotlightKid) [inline redbeat/schedulers.py]: Well, whaddayaknow, support for `decode_responses` was just merged into `fakeredis`. Still, I think it#s better, do decode explicitly.

Comment (sibson) [inline redbeat/schedulers.py]: Doing the decode inline with application logic makes it harder to understand the intent of the code.  I feel we should either use decode_responses or abstract out the decode into another method. Being lazy I'd opt for decode_reponses but am not opposed to doing it explicitly.

Comment (SpotlightKid) [inline redbeat/schedulers.py]: This means I need to do another PR introducing `decode_responses`, since some of the Python 3 compatibility fixes need to be adapted, and the `fakeredis` version needs to be pinned to `>=0.8.1` in the requirements.

**Comment by SpotlightKid** [comment]
I created a PR for using `decode_repsonses` (#39) and rebased this PR on it.

Hopefully, this can now be merged soon, and a new release issued, so we can run an official distribution in our production environment instead from our custom branch.

File: redbeat/schedulers.py [hunk 3]
```diff
@@ -189,23 +189,27 @@ def __init__(self, app, **kwargs):
         super(RedBeatScheduler, self).__init__(app, **kwargs)
 
     def setup_schedule(self):
-        # cleanup old static entries
+        # cleanup old static schedule entries
         client = redis(self.app)
-        previous = client.smembers(self.app.conf.REDBEAT_STATICS_KEY)
-        current = set(self.app.conf.CELERYBEAT_SCHEDULE.keys())
-        removed = previous - current
+        previous = set(key for key in client.smembers(self.app.conf.REDBEAT_STATICS_KEY))
```
Comment (sibson) [inline redbeat/schedulers.py]: is this comprehension necessary?

File: tests/test_scheduler.py [hunk 4]
```diff
@@ -52,3 +52,23 @@ def test_schedule_includes_current_and_next(self):
...
+        conf.CELERYBEAT_SCHEDULE = {
+            'test': {
+                'task': 'test',
+                'schedule': mocked_schedule(42)
+            }
+        }
+        s = self.create_scheduler()
+        redis = self.app.redbeat_redis
+
+        self.assertIn('test', s.schedule)
```
Comment (sibson) [inline tests/test_scheduler.py]: minor nit, using test twice impacts readability.

**Comment by sibson** [comment]
thanks for sticking with this and updating it.

File: redbeat/schedulers.py (hunk 3 above)
Comment (SpotlightKid) [inline redbeat/schedulers.py]: No. ;) There used to be a `key.decode()` here, which isn't needed anymore. Could be written like this now:

    previous = set(client.smembers(self.app.conf.REDBEAT_STATICS_KEY))
